atom/browser/native_window_mac.mm

```diff
@@ -95,6 +95,33 @@ - (void)windowDidResignMain:(NSNotification*)notification {
   shell_->NotifyWindowBlur();
 }
 
+- (NSSize)windowWillResize:(NSWindow *)sender toSize:(NSSize)frameSize {
+  NSSize newSize = frameSize;
+  double interiorContentAspectRatio = shell_->GetInteriorContentAspectRatio();
+
+  if (interiorContentAspectRatio > 0.0) {
+    NSRect windowFrame = [sender frame];
+    NSRect contentViewBounds = [[sender contentView] bounds];
+    gfx::Size interiorContentExtraSize = shell_->GetInteriorContentExtraSize();
+    double extraWidthPlusFrame = windowFrame.size.width - contentViewBounds.size.width + interiorContentExtraSize.width();
+    double extraHeightPlusFrame = windowFrame.size.height - contentViewBounds.size.height + interiorContentExtraSize.height();
+
+    newSize.width = roundf(((frameSize.height - extraHeightPlusFrame) * interiorContentAspectRatio) + extraWidthPlusFrame);
+
+    // If the new width is less than the frame size use it as the primary constraint. This ensures that the value returned
+    // by this method will never be larger than the users requested window size.
+    if (newSize.width < frameSize.width) {
+        newSize.height = roundf(((newSize.width - extraWidthPlusFrame) / interiorContentAspectRatio) + extraHeightPlusFrame);
```

Comment:
That doesn't get you a complete set of resizing behavior that is independent of the requested resize dimensions. Consider these scenarios: (with the bug fix mentioned at the end)

Scenario 1: (constrain the resize of height)

    frameSize = { 1000.0, 700.0 };
    aspectRatio = 2.0; // 2:1
    extraWidthPF = 0.0;
    extraHeightPF = 0.0;

    newSize.width = 1400.0 = ((700.0 - 0.0) * 2.0) + 0.0; // line 109
    if evaluates as false // line 113
    newSize.height = 500.0 = ((1000.0 - 0.0) / 2.0) + 0.0; // line 117
    newSize.width = 1000.0 = ((500.0 - 0.0) * 2.0) + 0.0; // line 118

    result of newSize = { 1000.0, 500.0 } // correct

Scenario 2: (constrain the resize of width)

    frameSize = { 1200.0, 500.0 };
    aspectRatio = 2.0; // 2:1
    extraWidthPF = 0.0;
    extraHeightPF = 0.0;

    newSize.width = 1000.0 = ((500.0 - 0.0) * 2.0) + 0.0; // line 109
    if evaluates as true // line 113
    newSize.height = 500.0 = ((1000.0 - 0.0) / 2.0) + 0.0; // line 114

    result of newSize = { 1000.0, 500.0 } // correct

Scenario 3: (constrain the height to within a resized width)

    frameSize = { 1300.0, 600.0 };
    aspectRatio = 2.0; // 2:1
    extraWidthPF = 0.0;
    extraHeightPF = 0.0;

    newSize.width = 1200.0 = ((600.0 - 0.0) * 2.0) + 0.0; // line 109
    if evaluates as true // line 113
    newSize.height = 600.0 = ((1200.0 - 0.0) / 2.0) + 0.0; // line 114

    result of newSize = { 1000.0, 600.0 } // correct

Scenario 4: (constrain the width to within a resized height)

    frameSize = { 1200.0, 700.0 };
    aspectRatio = 2.0; // 2:1
    extraWidthPF = 0.0;
    extraHeightPF = 0.0;

    newSize.width = 1400.0 = ((700.0 - 0.0) * 2.0) + 0.0; // line 109
    if evaluates as false // line 113
    newSize.height = 600.0 = ((1200.0 - 0.0) / 2.0) + 0.0; // line 117
    newSize.width = 1200.0 = ((600.0 - 0.0) * 2.0) + 0.0; // line 118

    result of newSize = { 1200.0, 600.0 } // correct

The goal is for the size of the result window to not be larger than the request. However we also want it to fill as much of the requested rectangle as possible.

Having me work through the exercise did help me discover a subtle bug. The comparison on line  113 should be less than or equal to which will be fixed when I check in the updated code.